### DISCUSSION TITLE
How to configure regexManager to update Go packages?

### DISCUSSION THREAD

**Discussion by karlhepler** [renovatebot/renovate#17797] [Request Help]
I have some bash scripts that install Go packages outside of the `go.mod` file definition. The intent is that these are self-installing developer tools.

They look something like this.
_There is some magic going on behind the scenes, which shouldn't be relevant here._

```
#!/usr/bin/env sh
dir=$(CDPATH='' cd -- "$(dirname -- "$0")" && pwd -P)

version='v0.0.0-20220822182546-3f2b496c77aa' # https://pkg.go.dev/golang.org/x/pkgsite/cmd/pkgsite?tab=versions
pkgpath='golang.org/x/pkgsite/cmd/pkgsite'

exec "$dir/.cacherun-goinstall" "$(basename "$0")" "$pkgpath" "$version" "$@"
```

I have similar scripts that I have successfully gotten to update GitHub releases.
Here's an example.

```
#!/usr/bin/env sh

set -e
[ "$DEBUG" = true ] && set -x

# renovate: datasource=github-releases depName=BurntSushi/ripgrep versioning=semver
version='13.0.0' # https://github.com/BurntSushi/ripgrep/releases

...
```

I'm trying to figure out how to use a `# renovate` comment to update the version in my first example. What should the versioning be? All I could find was [this document](https://docs.renovatebot.com/golang/#how-it-works) talking about how renovate updates Go modules.

**Comment by rarkins** [comment]
I'm not sure if using `github-tags` datasource or `go` datasource will be better for you. If you're using v0.0.0 versions then it would likely need to be `go`. If so then you'll need the packageName to be the fully qualified github package like would be extracted from a `go.mod` file. Versioning can hopefully just be the default `semver`.

**Reply by karlhepler** [reply]
I tried playing around with this solution, but I still wasn't able to get it to work.

Here's what I ended up with (still not working).

```
# renovate: datasource=go depName=golang.org/x/pkgsite versioning=semver-coerced
version='v0.0.0-20220913211724-eb1acd181321' # https://pkg.go.dev/golang.org/x/pkgsite/cmd/pkgsite?tab=versions
```

It's trying to fetch from GitHub for some reason and it can't find anything there.

I was able to discover the following information:

- The version seems to be in the following format: `v0.0.0-<committer date and time>-<first 12 of commit sha>`.
- The official git url of this package and version is here: https://go.googlesource.com/pkgsite/+/eb1acd1813213571638e55dc6e54bba83a92573b
- There is a way to list all versions in JSON format, here: https://go.googlesource.com/pkgsite/+log?format=JSON

Do you think there is any way I can get this working from all of that information?

**Reply by rarkins** [reply]
Can you create a minimal reproduction on GitHub.com?

**Reply by karlhepler** [reply]
Sure! Here you go! I went ahead and set up renovate with the repo as well. https://github.com/karlhepler/renovate-test

**Reply by rarkins** [reply]
Can you add a `go.mod` for the same dependency/value into this repo so that we can compare side by side?

**Reply by karlhepler** [reply]
Sure. I added a `go.mod` file and `main.go` file. I also committed the generated `go.sum` file. Interesting that the regular `go.mod` renovate feature does not detect the package. You can see that on the [dashboard](https://github.com/karlhepler/renovate-test/issues/1).

**Reply by karlhepler** [reply]
Maybe it's a problem with how they're releasing it? I believe it's meant be installed with `go install` instead of `go get`, since it's primary an executable with no shared Go packages. I think that's the intent, but I could be wrong.

**Reply by rarkins** [reply]
Unless Renovate's `gomod` manager detects it, it doesn't really help move this forward. Why did you make it `indirect`?

**Reply by karlhepler** [reply]
I didn't make it indirect. I ran 'go get'. The Go package manager did the rest.

**Reply by rarkins** [reply]
If you're then out of ideas, unfortunately so am I

**Reply by karlhepler** [reply]
Ok thanks.

**Reply by rarkins** [reply]
Just to summarize, we have hit a dead end due to inability to create a go project containing the same dependency, which Renovate recognizes. This is most likely due to the `// indirect` part. If that changes, then we can resume.

**Reply by karlhepler** [reply]
I was able to find a package that exports something and I imported that into my `main.go` file. I then ran `go mod tidy` and the `//indirect` comment disappeared. I now see that Renovate has picked it up. Please see [here](https://github.com/karlhepler/renovate-test/pull/2).

Given that, do you have any thoughts about how I might configure renovate to do the same for [test.bash](https://github.com/karlhepler/renovate-test/blob/main/test.bash)?

**Reply by rarkins** [reply]
Comparing them, gomod:

```json
               {
                 "managerData": {"lineNumber": 4},
                 "depName": "golang.org/x/pkgsite",
                 "depType": "require",
                 "currentValue": "v0.0.0-20220921205104-d06df0691bf0",
                 "datasource": "go",
                 "currentDigest": "d06df0691bf0",
                 "digestOneAndOnly": true,
                 "depIndex": 1,
                 "updates": [
                   {
                     "updateType": "digest",
                     "newValue": "v0.0.0-20220921205104-d06df0691bf0",
                     "newDigest": "86a1744034510e328a4ff3f90920efd7d523f584",
                     "branchName": "renovate/golang.org-x-pkgsite-digest"
                   }
                 ],
                 "warnings": [],
                 "versioning": "semver",
                 "sourceUrl": "https://github.com/golang/pkgsite",
                 "currentVersion": "v0.0.0-20220921205104-d06df0691bf0",
                 "fixedVersion": "v0.0.0-20220921205104-d06df0691bf0"
               }
```

regex:

```json
               {
                 "depName": "golang.org/x/pkgsite",
                 "currentValue": "v0.0.0-20220921205104-d06df0691bf0",
                 "datasource": "go",
                 "versioning": "semver-coerced",
                 "replaceString": "datasource=go depName=golang.org/x/pkgsite versioning=semver-coerced\nversion='v0.0.0-20220921205104-d06df0691bf0'\n",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "sourceUrl": "https://github.com/golang/pkgsite"
               }
```

I checked that `semver` vs `semver-coerced` wasn't the problem or at least the only problem, and it seems like to update a `v0.0.0` dependency we need `currentDigest` set, which isn't current possible (@viceice @secustor please correct me if I'm wrong). We might also need the `digestOneAndOnly` setting from `go` which is a bit of a hack. Unfortunately no good answer here so far!

BTW it might also be the case that >v1.0.0 packages *also* need `currentDigest`, but I'm not sure.

**Reply by karlhepler** [reply]
Thank you for looking into this.

**Comment by karlhepler** [comment]
Thanks. I’ll give it a try.
